python/dolma/core/utils.py

```diff
@@ -42,10 +41,11 @@ def split_paragraphs(text: str) -> List[TextSlice]:
     Split a string into paragraphs. A paragraph is defined as a sequence of zero or more characters, followed
     by a newline character, or a sequence of one or more characters, followed by the end of the string.
     """
-    return [
-        TextSlice(doc=text, start=match.start(), end=match.end())
+    paras = [
+        TextSlice(doc=text, start=match.start(), end=match.end() - 1)
```

Comment:
we need to keep `match.end()` because the text slice must contain `\n`, otherwise if any of the filters removes this paragraph it will leave dangling new lines.